internal/servers/controller/controller.go

```diff
@@ -123,6 +123,32 @@ func New(ctx context.Context, conf *Config) (*Controller, error) {
 		}
 	}
 
+	clusterListeners := make([]*base.ServerListener, 0)
+	for i := range conf.Listeners {
+		l := conf.Listeners[i]
+		if l == nil || l.Config == nil || l.Config.Purpose == nil {
+			continue
+		}
+		if len(l.Config.Purpose) != 1 {
+			return nil, fmt.Errorf("found listener with multiple purposes %q", strings.Join(l.Config.Purpose, ","))
```

Comment:
For my own understanding, why does this type allow multiple purposes if it is invalid here?

Reply:
The original plan was to support multiple purposes so that you could have a single port open and handle client, cluster, and API traffic. It's actually still possible but there are some gotchas and we just didn't have time to work through them before. The vestiges of this are in the alpnmux library that lives in the boundary repo.

The main place it gets complicated is if we end up allowing people to assign their own certs to workers' client bits, which may be needed in the future if we want to allow browser-based connections, since we can't control the TLS stack in a browser -- even when using wasm. This also means we can't control the ALPN value being sent by the browser, it is always `h2`, so if you have an API and worker client handler on the same port it becomes more complicated to deal with. Not impossible, but more complicated.

So for the moment we just have different ports in use. It'd be pretty trivial to allow cluster and API on the same port though.